maps.Rmd

```diff
@@ -299,52 +299,140 @@ ggplot(dawson[-69]) +
 
 <!-- HW: do you have any thoughts on a modern approach to this problem? -->
 
-Instead of displaying context with vector boundaries, you might want to draw a traditional map underneath. This is called a raster image. The easiest way to get a raster map of a given area is to use the ggmap package, which allows you to get data from a variety of online mapping sources including OpenStreetMap and Google Maps. Downloading the raster data is often time consuming so it's a good idea to cache it in a rds file, as illustrated here: \index{ggmap} \index{Raster data}
+A second way to supply geospatial information for mapping is to rely on **raster data**. Unlike the simple features format, in which geographical entities are specified in terms of a set of lines, points and polygons, rasters take the form of images. In the simplest case raster data might be nothing more than a bitmap file, but there are many different image formats out there. In the geospatial context specifically, there are image formats that include metadata (e.g., geodetic datum, coordinate reference system) that can be used to map the image information to the surface of the Earth. For example, one common format is GeoTIFF, which is a regular TIFF file with additional metadata supplied. Happily, most formats can be easily read into R with the assistance of GDAL (the Geospatial Data Abstraction Library, https://gdal.org/). For example the sf package contains a function `sf::gdal_read()` that provides access to the GDAL raster drivers from R. However, you rarely need to call this function directly, as there are other high level functions that take care of this for you. 
+
+As an illustration, suppose we wish to plot satellite images made publicly available by the Australian Bureau of Meterorology (BOM) on their FTP server. The bomrang package [@bomrang] provides a convenient interface to the server, including a `get_available_imagery()` function that returns a vector of filenames and a `get_satellite_imagery()` function that downloads a file and imports it directly into R. For expository purposes, however, I'll use a more flexible method that could be adapted to any FTP server, and use the `curl_download()` function from the curl package [@curl]
+
+```{r eval=FALSE}
+# list of all file names with time stamp 2020-01-07 21:00 GMT 
+# (BOM images are retained for 24 hours, so this will now return an
+# empty vector)
+files <- bomrang::get_available_imagery() %>%
+  stringr::str_subset("202001072100")
+
+# use curl_download() to obtain a single file, and purrr to 
+# vectorise this operation
+purrr::walk2(
+  .x = paste0("ftp://ftp.bom.gov.au/anon/gen/gms/", files),
+  .y = file.path("raster", files),
+  .f = ~ curl::curl_download(url = .x, destfile = .y)
+)
+```
+
+After caching the files locally (which is generally a good idea) we can inspect the list of files we have downloaded:
+
+```{r}
+list.files("raster")
```

Comment:
I prefer to use `dir()` since it's shorter

Reply:
HOW DID I NOT KNOW THERE WAS A DIR FUNCTION :facepalm: